lang/elaborator/src/normalizer/env.rs

```diff
@@ -8,56 +8,55 @@ use syntax::ast::{Shift, ShiftRange};
 use crate::normalizer::val::*;
 use printer::tokens::COMMA;
 use printer::Print;
-use syntax::ast::{Idx, Lvl, Var};
-use syntax::ctx::map_idx::*;
+use syntax::ast::{Idx, Var};
 use syntax::ctx::values::TypeCtx;
+use syntax::ctx::{map_idx::*, GenericCtx};
 use syntax::ctx::{Context, ContextElem, LevelCtx};
 
 #[derive(Debug, Clone, Derivative)]
 #[derivative(Eq, PartialEq, Hash)]
 pub struct Env {
-    bound: Vec<Vec<Rc<Val>>>,
+    ctx: GenericCtx<Rc<Val>>,
+}
+
+impl From<GenericCtx<Rc<Val>>> for Env {
+    fn from(value: GenericCtx<Rc<Val>>) -> Self {
+        Env { ctx: value }
+    }
 }
 
 impl Context for Env {
     type Elem = Rc<Val>;
 
     fn lookup<V: Into<Var>>(&self, idx: V) -> Self::Elem {
-        let lvl = self.var_to_lvl(idx.into());
-        self.bound
+        let lvl = self.ctx.var_to_lvl(idx.into());
+        self.ctx
+            .bound
             .get(lvl.fst)
             .and_then(|ctx| ctx.get(lvl.snd))
             .unwrap_or_else(|| panic!("Unbound variable {lvl}"))
             .clone()
     }
 
     fn push_telescope(&mut self) {
-        self.bound.push(vec![]);
+        self.ctx.bound.push(vec![]);
     }
 
     fn pop_telescope(&mut self) {
-        self.bound.pop().unwrap();
+        self.ctx.bound.pop().unwrap();
     }
 
     fn push_binder(&mut self, elem: Self::Elem) {
-        self.bound.last_mut().expect("Cannot push without calling level_inc_fst first").push(elem);
+        self.ctx
+            .bound
+            .last_mut()
+            .expect("Cannot push without calling level_inc_fst first")
```

Comment:
I think `level_inc_fst` here is outdated.

Reply:
Yes, it probably should be `push_telescope`. Changed in [eaebbdb](https://github.com/polarity-lang/polarity/pull/264/commits/eaebbdb50d46591306ed66e1c95c1e26f1db0970)